pkg/controller/upgrader/upgrader.go

```diff
@@ -0,0 +1,278 @@
+	client    *kubernetes.Clientset
+	config    *rest.Config
+	scheme    *runtime.Scheme
+	namespace string
+}
+
+func New(mgr manager.Manager) (*Upgrader, error) {
+	client, err := kubernetes.NewForConfig(mgr.GetConfig())
+	if err != nil {
+		log.Error(err, "Error creating kubernetes client")
+		return nil, err
+	}
+	return &Upgrader{
+		client:    client,
+		config:    mgr.GetConfig(),
+		scheme:    mgr.GetScheme(),
+		namespace: util.GetEnvOrDefault("NAMESPACE", "enmasse-infra"),
+	}, nil
+}
+
+func (u *Upgrader) Upgrade() error {
+
+	deploymentClient := u.client.AppsV1().Deployments(u.namespace)
+
+	var deployment *appsv1.Deployment
+	deployment, err := deploymentClient.Get(ADDRESS_SPACE_CONTROLLER_NAME, metav1.GetOptions{})
+	if err != nil && k8errors.IsNotFound(err) {
+		return nil
+	} else if err != nil {
+		return err
+	} else {
+		if _, ok := deployment.Annotations[ANNOTATION_VERSION]; !ok {
+			log.Info("address-space-controller is missing annotation, initiating upgrade")
+			err = u.performUpgrade(deployment)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (u *Upgrader) performUpgrade(addressSpaceControllerDeployment *appsv1.Deployment) error {
+	deploymentClient := u.client.AppsV1().Deployments(u.namespace)
+
+	// Scale down iot-operator
+	err := u.scale("iot-operator", 0)
+	if err != nil {
+		return err
+	}
+
+	// Scale down api-server
+	err = u.scale("api-server", 0)
+	if err != nil {
+		return err
+	}
+
+	// Scale down address-space-controller
+	err = u.scaleDeployment(addressSpaceControllerDeployment, 0)
+	if err != nil {
+		return err
+	}
+
+	// Scale down admin pods
+	admins, err := deploymentClient.List(metav1.ListOptions{
+		LabelSelector: "name=admin",
+	})
+	if err != nil {
+		return err
+	}
+	for _, adminDeployment := range admins.Items {
+		err = u.scaleDeployment(&adminDeployment, 0)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Scale down agent pods
+	agents, err := deploymentClient.List(metav1.ListOptions{
+		LabelSelector: "role=agent",
+	})
+	if err != nil {
+		return err
+	}
+	for _, agentDeployment := range agents.Items {
+		err = u.scaleDeployment(&agentDeployment, 0)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Get rid of API service so that it doesn't prevent upgrade
+	aggClient, err := aggregatorclient.NewForConfig(u.config)
+	if err != nil {
+		return err
+	}
+	apiServiceClient := aggClient.ApiregistrationV1().APIServices()
+
+	log.Info("Deleting v1beta1.enmasse.io api service")
+	err = deleteApiServiceIfPresent(apiServiceClient, "v1beta1.enmasse.io")
+	if err != nil {
+		return err
+	}
+
+	log.Info("Deleting v1alpha1.enmasse.io api service")
+	err = deleteApiServiceIfPresent(apiServiceClient, "v1alpha1.enmasse.io")
+	if err != nil {
+		return err
+	}
+
+	// Scale api-server back up
+	err = u.scale("api-server", 1)
+	if err != nil {
+		return err
+	}
+
+	// Scale address-space-controller back up
+	err = deployer.ApplyDeployment(addressSpaceControllerDeployment)
```

Comment:
This will set the annotation, and this will prevent the rest of the code in this function from being executing when the operator fails now.

I think this must be the last operation in this function.

Reply:
As discussed, this scenario is handled by the iotconfig-controller which will scale it back up in case it failed here.